### PR TITLE
Append warnings extracted before tail call execution

### LINKED ISSUE #6765: Warnings getting lost when returning a result of a call in tail position
Consider the following program:

```
from Standard.Base import all

do_select x =
    Warning.attach "WARN" [x+1, 2, 3]

do_fold_1 x =
    v = do_select x
    v.fold 0 (+)

do_fold_2 x =
    v = do_select x
    res = v.fold 0 (+)
    res

main =
    r1 = do_fold_1 100
    IO.println r1
    IO.println (Warning.get_all r1 . map .value)

    r2 = do_fold_2 100
    IO.println r2
    IO.println (Warning.get_all r2 . map .value)
```

### Expected Result

```
106
[WARN]
106
[WARN]
```

### Actual Result

```
106
[]
106
[WARN]
```

The methods `do_fold_1` and `do_fold_2` are basically the same. They do the same logic, just one of them uses a temporary variable. They should behave exactly the same. Yet, `do_fold_1` is losing warnings.

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

Throwing `TailCallException` meant that exceptions that were extracted from the expression before the call was made could not be appended. This change catches the `TailCallException`, adds warnings to it and propagates it further, thus ensuring that we don't loose the information.

Closes #6765.

### Important Notes

Removed workarounds introduced in stdlib.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/LoopingCallOptimiserNode.java [hunk 1]
```diff
@@ -79,13 +82,22 @@ public Object uncachedDispatch(
...
+        } else {
+          return result;
+        }
       } catch (TailCallException e) {
         function = e.getFunction();
         callerInfo = e.getCallerInfo();
         arguments = e.getArguments();
+        if (warnings == null) {
+          warnings = e.getWarnings();
+        }
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/LoopingCallOptimiserNode.java]: What if `warnings != null`?

Shouldn't we somehow concatenate the existing warnings with the new ones?

**Review by radeusgd — APPROVED** [review]
Thanks for removing the outdated workarounds. The tests look all good.

I just have one question as I'm not sure if I correctly understand a piece of logic. Other than that - looks good.

**Review by 4e6 — APPROVED** [review]
Every time I look at the engine internals and realize how much bookkeeping is done behind the scenes. I genuinely wonder how it keeps maintaining a decent performance :smile:

**Review by GregoryTravis — APPROVED** [review]

**Comment by hubertp** [comment]
> Every time I look at the engine internals and realize how much bookkeeping is done behind the scenes. I genuinely wonder how it keeps maintaining a decent performance smile

Warnings affect performance. Quite significantly TBH (benchmarks show it).

**Comment by 4e6** [comment]
sad panda

**Review by Akirathan — APPROVED** [review]

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
This is not what I envisioned. I don't understand why TCE should have `Warnings` field.

I wrote: 
> How such "loosing of warnings" can happen?  `v.fold 0 (+)` in `do_fold_1` strips warnings off and calls `fold` on "real v" value.
> `fold` uses `@Tail_Call` and then it returns. The stripped off warnings aren't added.
> But how do you want to carry warnings in `TailCallException`? They are already stripped off when it is created!
> The fix is: When dealing with `Warnings`, catch `TailCallException` and re-throw with warnings re-attached.

**Comment by JaroslavTulach** [comment]
What are the benchmark results? Is regular TCO slower?

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java [hunk 2]
```diff
@@ -339,8 +340,12 @@ Object doWarning(
 
     arguments[thisArgumentPosition] = selfWithoutWarnings;
 
-    Object result = childDispatch.execute(frame, state, symbol, selfWithoutWarnings, arguments);
-    return WithWarnings.appendTo(EnsoContext.get(this), result, arrOfWarnings);
+    try {
+      Object result = childDispatch.execute(frame, state, symbol, selfWithoutWarnings, arguments);
+      return WithWarnings.appendTo(EnsoContext.get(this), result, arrOfWarnings);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: There is following code above this change:
```java
    if (childDispatch == null) {
      CompilerDirectives.transferToInterpreterAndInvalidate();
      Lock lock = getLock();
      lock.lock();
      try {
        if (childDispatch == null) {
          childDispatch =
              insert(
                  build(
                      invokeFunctionNode.getSchema(),
                      invokeFunctionNode.getDefaultsExecutionMode(),
                      invokeFunctionNode.getArgumentsExecutionMode(),
                      thisArgumentPosition));
          childDispatch.setTailStatus(getTailStatus());
          childDispatch.setId(invokeFunctionNode.getId());
          notifyInserted(childDispatch);
        }
      } finally {
        lock.unlock();
      }
    }
```

1. it is wrong: we should create a factory method and use `@Cached` - I don't think there is a reason for that not to work
2. it sets `setTailStatus`

Shouldn't we have two `@Specialization`s - one for values `WithWarnings` and one for without warnings and just make sure the `TailStatus` for `WithWarnings` is `NOT_TAIL`?

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/CurryNode.java [hunk 3]
```diff
@@ -153,7 +153,8 @@ private Object doCall(
       VirtualFrame frame, Function function, CallerInfo callerInfo, State state, Object[] arguments) {
     return switch (getTailStatus()) {
       case TAIL_DIRECT -> directCall.executeCall(frame, function, callerInfo, state, arguments);
-      case TAIL_LOOP -> throw new TailCallException(function, callerInfo, arguments);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/CurryNode.java]: Unnecessary change.

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/LoopingCallOptimiserNode.java [hunk 4]
```diff
@@ -79,13 +82,22 @@ public Object uncachedDispatch(
       State state,
       Object[] arguments,
       @Cached ExecuteCallNode executeCallNode) {
+    Warning[] warnings = null;
     while (true) {
       try {
-        return executeCallNode.executeCall(frame, function, callerInfo, state, arguments);
+        Object result = executeCallNode.executeCall(frame, function, callerInfo, state, arguments);
+        if (warnings != null) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/LoopingCallOptimiserNode.java]: I am fine that usage of `Warnings` slows execution down, however in this case we are slowing _everything_ down - be there warnings or not, this change has to be made. I'd like us to find another solution.

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/SimpleCallOptimiserNode.java [hunk 5]
```diff
@@ -65,8 +67,13 @@ public Object executeDispatch(
           lock.unlock();
         }
       }
-      return next.executeDispatch(
-          frame, e.getFunction(), e.getCallerInfo(), state, e.getArguments());
+      Object result =
+          next.executeDispatch(frame, e.getFunction(), e.getCallerInfo(), state, e.getArguments());
+      if (e.getWarnings() != null) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/SimpleCallOptimiserNode.java]: Again, this would affect all executions, even those that don't use warnings, I am afraid.

**Comment by hubertp** [comment]
> What are the benchmark results? 

https://enso-org.github.io/engine-benchmark-results/
The last ones show it - Vectors w/ vs w/o warnings. I remember us talking about it and saying that this will need a bit more work but it is not a priority. 

> Is regular TCO slower?

No

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/CurryNode.java (hunk 3 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/CurryNode.java]: Typo, yes

**Comment by hubertp** [comment]
> I wrote:
> 
> > How such "loosing of warnings" can happen?  `v.fold 0 (+)` in `do_fold_1` strips warnings off and calls `fold` on "real v" value.
> > `fold` uses `@Tail_Call` and then it returns. The stripped off warnings aren't added.
> > But how do you want to carry warnings in `TailCallException`? They are already stripped off when it is created!
> > The fix is: When dealing with `Warnings`, catch `TailCallException` and re-throw with warnings re-attached.

Can't re-attach warnings to the result of method call because there is no result to attach to at that point in time. That's why I enhanced TCE to carry that info when the result is available.

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java (hunk 2 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: I can explore the solution with another specialization.

Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: > one for values WithWarnings and one for without warnings

This specialization has `hasWarnings(self)` guard so the second option is not possible.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: I don't get it. What is not possible? The signature of the method is:
```java
...
 @CachedLibrary(limit = "10") WarningsLibrary warnings) {
```
I believe it should be changed to:
```diff
...
diff --git engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
index 7949001e78..1d3c71c769 100644
--- engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
+++ engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
@@ -15,6 +15,11 @@ import com.oracle.truffle.api.nodes.RootNode;
 import com.oracle.truffle.api.profiles.BranchProfile;
 import com.oracle.truffle.api.profiles.ConditionProfile;
 import com.oracle.truffle.api.source.SourceSection;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.UUID;
 import org.enso.interpreter.node.BaseNode;
 import org.enso.interpreter.node.MethodRootNode;
 import org.enso.interpreter.node.callable.dispatch.InvokeFunctionNode;
@@ -44,19 +49,11 @@ import org.enso.interpreter.runtime.error.WithWarnings;
 import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
 import org.enso.interpreter.runtime.state.State;
 
-import java.time.LocalDate;
-import java.time.LocalTime;
-import java.time.ZoneId;
-import java.time.ZonedDateTime;
-import java.util.UUID;
-import java.util.concurrent.locks.Lock;
-
 @ImportStatic({HostMethodCallNode.PolyglotCallType.class, HostMethodCallNode.class})
 public abstract class InvokeMethodNode extends BaseNode {
   protected static final int CACHE_SIZE = 10;
   private @Child InvokeFunctionNode invokeFunctionNode;
   private final ConditionProfile errorReceiverProfile = ConditionProfile.createCountingProfile();
-  private @Child InvokeMethodNode childDispatch;
 
   private final int argumentCount;
   private final int thisArgumentPosition;
@@ -97,8 +94,10 @@ public abstract class InvokeMethodNode extends BaseNode {
   public void setTailStatus(TailStatus tailStatus) {
     super.setTailStatus(tailStatus);
     this.invokeFunctionNode.setTailStatus(tailStatus);
-    if (childDispatch != null) {
-      childDispatch.setTailStatus(tailStatus);
+    for (Node n : getChildren()) {
+      if (n instanceof InvokeMethodNode childDispatch) {
+        childDispatch.setTailStatus(tailStatus);
+      }
     }
   }
 
@@ -296,6 +295,18 @@ public abstract class InvokeMethodNode extends BaseNode {
     return warningFunctionNode.execute(resolvedFunction, frame, state, arguments1);
   }
 
+  final InvokeMethodNode buildChildDispatch() {
+    var childDispatch = build(
+            invokeFunctionNode.getSchema(),
+            invokeFunctionNode.getDefaultsExecutionMode(),
+            invokeFunctionNode.getArgumentsExecutionMode(),
+            thisArgumentPosition
+    );
+    childDispatch.setTailStatus(getTailStatus());
+    childDispatch.setId(invokeFunctionNode.getId());
+    return childDispatch;
+  }
+
   @Specialization(guards = "warnings.hasWarnings(self)")
   Object doWarning(
       VirtualFrame frame,
@@ -303,7 +314,9 @@ public abstract class InvokeMethodNode extends BaseNode {
       UnresolvedSymbol symbol,
       Object self,
       Object[] arguments,
-      @CachedLibrary(limit = "10") WarningsLibrary warnings) {
+      @CachedLibrary(limit = "10") WarningsLibrary warnings,
+      @Cached("buildChildDispatch()") InvokeMethodNode childDispatch
+  ) {
     Object selfWithoutWarnings;
     Warning[] arrOfWarnings;
     try {
@@ -313,30 +326,6 @@ public abstract class InvokeMethodNode extends BaseNode {
       // Can't throw `CompilerDirectives.shouldNotReachHere` as it crashes native-image build
       throw new IllegalStateException(e);
     }
-
-    // Cannot use @Cached for childDispatch, because we need to call notifyInserted.
-    if (childDispatch == null) {
-      CompilerDirectives.transferToInterpreterAndInvalidate();
-      Lock lock = getLock();
-      lock.lock();
-      try {
-        if (childDispatch == null) {
-          childDispatch =
-              insert(
-                  build(
-                      invokeFunctionNode.getSchema(),
-                      invokeFunctionNode.getDefaultsExecutionMode(),
-                      invokeFunctionNode.getArgumentsExecutionMode(),
-                      thisArgumentPosition));
-          childDispatch.setTailStatus(getTailStatus());
-          childDispatch.setId(invokeFunctionNode.getId());
-          notifyInserted(childDispatch);
-        }
-      } finally {
-        lock.unlock();
-      }
-    }
-
     arguments[thisArgumentPosition] = selfWithoutWarnings;
 
     Object result = childDispatch.execute(frame, state, symbol, selfWithoutWarnings, arguments);
@@ -720,8 +709,10 @@ public abstract class InvokeMethodNode extends BaseNode {
    */
   public void setId(UUID id) {
     invokeFunctionNode.setId(id);
-    if (childDispatch != null) {
-      childDispatch.setId(id);
+    for (var n : getChildren()) {
+      if (n instanceof InvokeMethodNode childDispatch) {
+        childDispatch.setId(id);
+      }
     }
   }
 }
```
The `for (var n : getChildren()) { if (n instanceof InvokeMethodNode childDispatch) {` isn't really nice, but we no longer have pointer to the `childDispatch` node. There could be a problem with instrumentation, but `InvokeMethodNode` isn't instrumentable. Things should continue to work the as before. The generated code looks good:
```java
WarningsLibrary warning_warnings__ = super.insert((WARNINGS_LIBRARY_.createDispatched(10)));
if ((warning_warnings__.hasWarnings(arg2Value))) {
  this.warning_warnings_ = super.insert(warning_warnings__);
  this.warning_childDispatch_ = super.insert((buildChildDispatch()));
  this.state_0_ = state_0 = state_0 | 0b100000 /* add-state_0 doWarning(VirtualFrame, State, UnresolvedSymbol, Object, Object[], WarningsLibrary, InvokeMethodNode) */;
  lock.unlock();
  hasLock = false;
  return doWarning(frameValue, arg0Value, arg1Value, arg2Value, arg3Value, warning_warnings__, this.warning_childDispatch_);
}
```


If you like the patch, feel inspired to apply it.

Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: My patch looks exactly the same. What I meant by it is that I initially understood your " one for values WithWarnings and one for without warnings " comment as we should split the existing `doWarning` into two cases. But based on your comment, this is not what you meant. A case of a message _lost in translation_.

> There could be a problem with instrumentation, but InvokeMethodNode isn't instrumentable. 

I believe that's the primary reason why we don't use `@Cached` in this and other places. See the 
```
// Cannot use @Cached for childDispatch, because we need to call notifyInserted.
```
comment.

**Comment by hubertp** [comment]
@JaroslavTulach Couldn't change to `@Cached` because of the instrumentation. There is a `WarningInstrumentatiopnTest` illustrating the problem.

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: `notifyInserted`! Not `insert`. I see.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: > we should create a factory method and use @Cached - I don't think there is a reason for that not to work

OK, so we know that it doesn't work because of `WarningInstrumentatiopnTest`. Sorry for false hope request.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: > based on your comment, this is not what you meant.

Didn't I? What is wrong with:

> Shouldn't we have two @Specializations - one for values WithWarnings and one for without warnings and just make sure the TailStatus for WithWarnings is NOT_TAIL?

It still solves the problem of checking warnings all the time. It even simplifies the code as there is no need to do anything special when there are warnings. Just not to allow tail calls at that moment, for such values.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: After [5628c71](https://github.com/enso-org/enso/pull/6849/commits/5628c711997ae3b0de0eb6f36e5ac4011660b15e) I see just three changes setting `NOT_TAIL`:
- [InvokeCallableNode](https://github.com/enso-org/enso/pull/6849/files#diff-01d58f3323980e95f4bf73d7c122e1633dafc4f85fa49661106b82755f0aeaa4R280)
- [InvokeConversionNode](https://github.com/enso-org/enso/pull/6849/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R331)
- [InvokeMethodNode](https://github.com/enso-org/enso/pull/6849/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R331)

That seems like the right fix.

**Review by JaroslavTulach — APPROVED** [review]
After [5628c71](https://github.com/enso-org/enso/pull/6849/commits/5628c711997ae3b0de0eb6f36e5ac4011660b15e) I see just three changes setting `NOT_TAIL`:
- [InvokeCallableNode](https://github.com/enso-org/enso/pull/6849/files#diff-01d58f3323980e95f4bf73d7c122e1633dafc4f85fa49661106b82755f0aeaa4R280)
- [InvokeConversionNode](https://github.com/enso-org/enso/pull/6849/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R331)
- [InvokeMethodNode](https://github.com/enso-org/enso/pull/6849/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R331)

That seems like the right fix.
